### PR TITLE
Fix for parsing incremental input

### LINKED ISSUE #11069: Unexpected behavior of the parser API
I can reproduce the issue with the C++ API. The issue was already present in v1.1.2.

### Discussed in https://github.com/cvc5/cvc5/discussions/11001

<div type='discussions-op-text'>

<sup>Originally posted by **Flower-Wang1024** July  3, 2024</sup>
```python
import cvc5

tm = cvc5.TermManager()
solver = cvc5.Solver(tm)
sm = cvc5.SymbolManager(solver)

solver.setLogic("QF_BV")

parser = cvc5.InputParser(solver, sm)

input1 = "(declare-const x (_ BitVec 4))\n" \
               "(assert (= x #b0001))\n"

parser.setIncrementalStringInput(cvc5.InputLanguage.SMT_LIB_2_6, "myInput")
parser.appendIncrementalStringInput(input1)

# while True:
#     cmd = parser.nextCommand()
#     if cmd.isNull():
#         break
#     cmd.invoke(solver, sm)
# result = solver.checkSat()
# print("Result:", result)

input2 = "(assert (= x #b0101))"
parser.appendIncrementalStringInput(input2)

while True:
    cmd = parser.nextCommand()
    if cmd.isNull():
         break
    cmd.invoke(solver, sm)

result = solver.checkSat()
print("Result:", result)
```
Hello! I'm a student who are learning. I met a problem while trying to solve smt2 incrementally. If I uncomment the code, why can I parse only input1?  Is there something wrong with my code?
</div>

### PR DESCRIPTION
**Pull request by ajreynol**
When using the incremental input interface, the parser had the behavior that if it was ever finished, it did not reset its "finished" flag even if more input was added.

This makes the parser internally reallocate a lexer and parser if the user had exhausted all their input.

Fixes #11069.

### CODE REVIEW

File: include/cvc5/cvc5_parser.h [hunk 1]
```diff
@@ -308,6 +308,11 @@ class CVC5_EXPORT InputParser
    */
   void setStringInputInternal(const std::string& input,
                               const std::string& name);
+  /**
+   * Set the input to incremental string input.
```
Comment (aniemetz) [inline include/cvc5/cvc5_parser.h:312]: Parameters are not documented.

File: include/cvc5/cvc5_parser.h [hunk 2]
```diff
@@ -325,6 +330,10 @@ class CVC5_EXPORT InputParser
   std::stringstream d_istringStream;
   /** Are we initialized to use the above string stream? */
   bool d_usingIStringStream;
+  /** The last language argument passed to setIncrementalStringInput */
+  modes::InputLanguage d_istringLang;
+  /** The last name arg passed to setIncrementalStringInput */
```
Comment (aniemetz) [inline include/cvc5/cvc5_parser.h:335]:
```suggestion
  /** The last name argument passed to setIncrementalStringInput(). */
```

File: include/cvc5/cvc5_parser.h [hunk 3]
```diff
@@ -325,6 +330,10 @@ class CVC5_EXPORT InputParser
   std::stringstream d_istringStream;
   /** Are we initialized to use the above string stream? */
   bool d_usingIStringStream;
+  /** The last language argument passed to setIncrementalStringInput */
```
Comment (aniemetz) [inline include/cvc5/cvc5_parser.h:333]:
```suggestion
  /** The last language argument passed to setIncrementalStringInput(). */
```
@ajreynol you have to use `()` for function references for Doxygen to pick this up properly and link to the function.

File: src/api/cpp/cvc5_parser.cpp [hunk 4]
```diff
@@ -345,23 +346,33 @@ void InputParser::setStringInputInternal(const std::string& input,
...
+void InputParser::setIncrementalStringInputInternal(modes::InputLanguage lang,
+                                                    const std::string& name)
+{
+  // initialize the parser
+  d_parser = Parser::mkParser(lang, d_solver, d_sm->toSymManager());
+  initializeInternal();
+  d_istringStream.str("");
+  d_istringStream.clear();
+  d_parser->setStreamInput(d_istringStream, name);
+}
```
Comment (aniemetz) [inline src/api/cpp/cvc5_parser.cpp:358]: This is missing the usual `CVC5_API_TRY_CATCH_BEGIN` handling.

**Review by aniemetz — CHANGES_REQUESTED** [review]
Can you also add a corresponding test for the C bindings in (new directory) `test/api/c`?

File: src/api/cpp/cvc5_parser.cpp (hunk 4 above)
Comment (ajreynol) [reply src/api/cpp/cvc5_parser.cpp:358]: This is an internal helper method, so this shouldn't be done, right?

Comment (aniemetz) [reply src/api/cpp/cvc5_parser.cpp:358]: Ah right, in this case it's fine.

File: include/cvc5/cvc5_parser.h [hunk 5]
```diff
@@ -308,6 +308,13 @@ class CVC5_EXPORT InputParser
    */
   void setStringInputInternal(const std::string& input,
                               const std::string& name);
+  /**
+   * Set the input to incremental string input.
+   * @param lang the input language
+   * @param name the name of the stream, for use in error messages
```
Comment (aniemetz) [inline include/cvc5/cvc5_parser.h:314]:
```suggestion
   * @param lang The input language.
   * @param name The name of the stream, for use in error messages.
```

**Comment by aniemetz** [comment]
#11095 sets up the C api test infrastructure, after this is merged in, some of the issues of this PR will be resolved.

The current API test should still be converted to a `.c` file (instead of a `.cpp` file).

File: test/api/c/issue11069.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  std::cout << "Result:" << result << std::endl;
+
+  std::string input2("(assert (= x #b0101))");
+  cvc5_parser_append_inc_str_input(parser, input2.c_str());
+  const char* error_msg;
+  cmd = cvc5_parser_next_command(parser, &error_msg);
+  assert(!cmd.isNull());
+  (void)cvc5_cmd_invoke(cmd, &slv, &sm);
+  result = cvc5_check_sat(d_solver);
+  std::cout << "Result:" << result << std::endl;
```
Comment (daniel-larraz) [inline test/api/c/issue11069.c:60]: The following changes fix the test for the C bindings:
```suggestion
#include <stdio.h>
#include <assert.h>

int main()
{
  Cvc5TermManager* tm = cvc5_term_manager_new();
  Cvc5* slv = cvc5_new(tm);

  Cvc5SymbolManager* sm = cvc5_symbol_manager_new(tm);

  cvc5_set_logic(slv, "QF_BV");

  // construct an input parser associated the solver above
  Cvc5InputParser* parser = cvc5_parser_new(slv, sm);

  const char* input1 = "(declare-const x (_ BitVec 4))\n(assert (= x #b0001))\n";
  cvc5_parser_set_inc_str_input(
          parser, CVC5_INPUT_LANGUAGE_SMT_LIB_2_6, "myInput");
  cvc5_parser_append_inc_str_input(parser, input1);

  // parse commands until finished
  Cvc5Command cmd;
  while (true)
  {
    const char* error_msg;
    cmd = cvc5_parser_next_command(parser, &error_msg);
    if (cmd == NULL)
    {
      break;
    }
    (void)cvc5_cmd_invoke(cmd, slv, sm);
  }
  Cvc5Result result = cvc5_check_sat(slv);
  printf("Result: %s\n", cvc5_result_to_string(result));

  const char* input2 = "(assert (= x #b0101))";
  cvc5_parser_append_inc_str_input(parser, input2);
  const char* error_msg;
  cmd = cvc5_parser_next_command(parser, &error_msg);
  assert(cmd != NULL);
  (void)cvc5_cmd_invoke(cmd, slv, sm);
  result = cvc5_check_sat(slv);
  printf("Result: %s\n", cvc5_result_to_string(result));
```

**Review by daniel-larraz — APPROVED** [review]

**Pull request merged**